evol/helpers/checkpointing.py

```diff
@@ -0,0 +1,59 @@
+import json
+import pickle
+from datetime import datetime
+from typing import List
+
+from os import listdir
+from os.path import isdir, exists, join
+
+from evol import Individual
+
+
+def checkpoint(individuals: List[Individual], directory: str, method: str='pickle') -> None:
+    """Checkpoint a list of individuals.
+
+    :param individuals: List of individuals to checkpoint.
+    :param directory: Location to store the checkpoint. A new file is created for every checkpoint.
+    :param method: One of "pickle" or "json". For json, the chromosomes need to be json-serializable.
+    """
+    if not isdir(directory):
+        raise FileNotFoundError('Cannot write to "{}": is not a directory.'.format(directory))
+    filename = join(directory,
+                    datetime.now().strftime("%Y%m%d-%H%M%S.%f") + ('.pkl' if method == 'pickle' else '.json'))
+    if exists(filename):
+        raise FileExistsError('Cannot write to "{}": file exists!'.format(filename))
+    if method == 'pickle':
+        with open(filename, 'wb') as pickle_file:
+            pickle.dump(individuals, pickle_file)
+    elif method == 'json':
+        with open(filename, 'w') as json_file:
+            json.dump([individual.__dict__ for individual in individuals], json_file)
+    else:
+        raise ValueError('Invalid checkpointing method "{}". Choose "pickle" or "json".'.format(method))
+
+
+def load(path: str) -> List[Individual]:
```

Comment:
So I've decided to create functions outside the implementation of `Population` as I felt it does not belong there. On the other hand, perhaps it shouldn't be helpers either. Should I move this to a separate `serialization.py` next to (not inside) `helpers`?

Reply:
That feels better.

Reply:
It does feel like the approach from `logging` is significantly different from `serialization` even though they are doing rather similar things. Wouldn't it be better to pass an object to the population that tells it where and how to serialise? 

This is *VincentThinkingOutLoud* but it could result into something like: 

```python
pop = Population(chromosomes=chromosomes, eval_function=eval_func, 
                 logger=BaseLogger("/tmp/logs/evol.log"), 
                 serializer=JsonSerializer("/tmp/chckpnts/"))

evo = Evolution().survive(...).breed(..., ...).log().checkpoint(every=5)
pop.repeat(evolution=evo, n=1000)
```

Good idea/ bad idea?

Reply:
I like the idea of having a single way to do things, but in this case I'm afraid it will complicate things. Logging is a lot like serialisation, but is different in two aspects:

- Logging may happen in many formats and to many types of destinations. I don't foresee any other formats than JSON and Pickle for serialisation, and I really don't foresee serialisation to other destinations than local disk.
- I expect many people to only think about serialisation once they have spent some time running their algorithm. I want to offer them a way to checkpoint even if they haven't initialised their Population with a serializer. When it comes to logging -- of course we can't log after the fact, but for checkpointing one can simply take the resulting population and use `.checkpoint` on it.

Reply:
I am reminded that we may want to add an `every` keyword to `.log()` as well. 

1. Thinking about it -> logging is usually to disk too. And there won't be any other format than text. In some edge cases when you are running this on many server I guess you may want to log to a distributed stream. But this is not the typical/beginner usecase. In that regard I would argue that logging and serliazing are actually very similar. 
2. You'd either pass in the serializing settings to the `Population()`-call or to the `.checkpoint` call either way. The question then becomes, should the UI be different? I mainly worry about confusion.

Reply:
Good call for the `every` on `log()`.

I understand your worries and party agree. The way I see it is this: logging and serialising are very similar, but on a few points also fundamentally different. The interface to both is rather similar: `log()` and `checkpoint()` can be called on `Population` and `Evolution`s. The big difference is that while with `checkpoint()` you want to store exactly the same thing at all times (it has to be the same such that you can read it again), with logging the _what_ and _how_ (what format) may vary. 

Currently the checkpointing expects the path in the `checkpoint()` method, while the logger expects the path upon initialisation. This I don't like. I suggest we provide the path to the `log()`, just like is the case with serialization. In the simplest use-case (where one uses the base logger, which I think will be basically always) the interface will then be exactly the same.

Reply:
Summarizing the discussion we had offline; we will initialise the `Population` with ` SimpleSerializer` which handles the checkpointing to pickle and json. Users are free to write others that serialise to/from weird alternatives.

The `Population` will accept a `checkpoint_path` argument, which will be the default path to checkpoint to. `checkpoint()` can be used without `path=` when the default is set. If `path=` is provided to `.checkpoint()`, this overrides the default.

We will try to do the same for logging (including renaming the logger to `SimpleLogger`). I suggest we wait with the changes to logging until the serialisation has the required form.